python/nav/statemon/db.py

```diff
@@ -246,11 +246,37 @@ def commit_event(self, event):
         values = (nextid, 'descr', event.info)
         self.execute(statement, values)
 
-    def hosts_to_ping(self):
-        """Returns a list of netboxes to ping, from the database"""
-        query = """SELECT netboxid, sysname, ip, up FROM netbox """
+    def hosts_to_ping(self, groups_included=None, groups_excluded=None):
+        """Returns a list of netboxes to ping, from the database
+
+        :param groups_included: A list of device group names whose members will be
+                                the only ones included in the result.
+        :type groups_included: list
+        :param groups_excluded: A list of device group names whose members will be
+                                excluded from the result.
+        :type groups_excluded: list
+        """
+
+        if groups_included is None:
+            groups_included = []
+        if groups_excluded is None:
+            groups_excluded = []
```

Comment:
Now we're just down to nitpicking here :)

While I agree with @hmpf that some tests would be nice, I don't necessarily agree that this function has grown too large to grok. It's still less than a page long. If I understand @hmpf correctly, she's proposing to test that the resulting query string matches something predefined - though I'm not sure. Maybe @hmpf could elaborate?

Things like these four lines are more or less redundant, given the logic below, and could be easily removed, decreasing the cognitive load of reading this code.

Reply:
So is something like this what you had in mind @hmpf ? 
```
    def build_host_query(self, groups_included=None, groups_excluded=None):
        """Returns a tuple of (query: str, query_parameters: list)

        :param groups_included: A list of device group names whose members will be
                                the only ones included in the result.
        :type groups_included: list
        :param groups_excluded: A list of device group names whose members will be
                                excluded from the result.
        :type groups_excluded: list
        """

        query = """SELECT distinct(netbox.netboxid), sysname, ip, up FROM netbox
                   LEFT JOIN netboxcategory USING (netboxid)"""

        params = []
        if groups_included:
            query += " WHERE netboxcategory.category IN %s"
            params.append(tuple(groups_included))

        if groups_excluded:
            query += " AND " if groups_included else " WHERE "
            query += "(netboxcategory.category IS NULL OR netboxcategory.category NOT IN %s)"
            params.append(tuple(groups_excluded))

        return query, params

    def hosts_to_ping(self, groups_included=None, groups_excluded=None):
        """Returns a list of netboxes to ping, from the database

        :param groups_included: A list of device group names whose members will be
                                the only ones included in the result.
        :type groups_included: list
        :param groups_excluded: A list of device group names whose members will be
                                excluded from the result.
        :type groups_excluded: list
        """

        query, params = self.build_host_query(groups_included, groups_excluded)

        try:
            self._hosts_to_ping = self.query(query, params)
        except DbError:
            return self._hosts_to_ping
        return self._hosts_to_ping
```

Reply:
Yes! One (mental) page of code for me is 24 chars high. that's about the max of what I can take in without reading =) VT220 damage no doubt.